### PR TITLE
[AOTAutograd] Fix CSE to deduplicate NaN constant tensors by normalizing float/complex hashing and comparison

### LINKED ISSUE #191013: torch.compile CSE does not deduplicate constant NaN tensors because they don't compare equal
### 🐛 Describe the bug

Take this repro:
```py
import torch

@torch.compile(backend="inductor", fullgraph=True)
def repeated_fills(x: torch.Tensor) -> tuple[torch.Tensor, ...]:
    finite_0 = torch.full_like(x, 1.0)
    finite_1 = torch.full_like(x, 1.0)
    nan_0 = torch.full_like(x, float("nan"))
    nan_1 = torch.full_like(x, float("nan"))
    return x + finite_0, x - finite_1, x + nan_0, x - nan_1

def main() -> None:
    # requires_grad=True selects torch.compile's training path, where AOTAutograd
    # runs its CSE pass before partitioning the joint forward-backward graph.
    x = torch.randn(8, requires_grad=True)
    outputs = repeated_fills(x)
    sum(output.sum() for output in outputs).backward()


if __name__ == "__main__":
    main()
```

Run it with `TORCH_LOGS=aot_graphs python repro.py`.

You'll see that this is the optimized forward graph:
```
full_default: "f32[8][1]cpu" = torch.ops.aten.full.default([8], 1.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
full_default_2: "f32[8][1]cpu" = torch.ops.aten.full.default([8], nan, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
full_default_3: "f32[8][1]cpu" = torch.ops.aten.full.default([8], nan, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
add: "f32[8][1]cpu" = torch.ops.aten.add.Tensor(primals_1, full_default)
sub: "f32[8][1]cpu" = torch.ops.aten.sub.Tensor(primals_1, full_default);  full_default = None
add_1: "f32[8][1]cpu" = torch.ops.aten.add.Tensor(primals_1, full_default_2);  full_default_2 = None
sub_1: "f32[8][1]cpu" = torch.ops.aten.sub.Tensor(primals_1, full_default_3);  primals_1 = full_default_3 = None
```

The two tensors filled with ones get deduplicated, the two tensors filled with NaNs do not.

This is causing problems in practice. In particular, it's preventing CSE when using torchao's mxfp8 recipe where a single tensor gets quantized multiple times by different consumers (e.g., the residual being fed to wq, wk and wv), because of this line: https://github.com/pytorch/ao/blob/693b573f6f3eae8be467db0ef3e49a1553ae2731/torchao/prototype/mx_formats/mx_tensor.py#L188

### Versions

PyTorch 2.13.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by KarhouTam**
Fixes #191013

## TL;DR
Fix CSE to deduplicate NaN constant tensors by normalizing float/complex hashing and comparison using their IEEE 754 bit patterns.

## Problem
The AOTAutograd CSE pass (`fx_graph_cse`) was failing to deduplicate identical constant tensors (e.g., multiple `torch.full_like(x, float("nan"))` calls) because Python float semantics break both the hash lookup and the token equality check used by CSE:

- NaN: `float("nan") != float("nan")` (IEEE 754), and `hash(float("nan"))` is id-based in CPython, so two NaN instances hash differently.
- Negative zero: `float(-0.0) == float(0.0)` (IEEE 754), so distinct constants incorrectly merge.

## Solution
The fix adds `_normalize_cse_arg()`, which replaces float and complex values with a `_ScalarKey` dataclass tagged by their IEEE 754 bit pattern (via `struct.pack`). This gives consistent hashing and equality for NaN values while preserving distinct bit patterns (NaN sign, negative zero).

A `_ScalarKey(tag, bits)` tuple is used rather than the raw bit pattern alone to avoid collisions between float and complex values that happen to have the same byte representation.

## Test Plan
```
python test/functorch/test_memory_efficient_fusion.py
```

New tests cover:
- NaN deduplication (identical +nan constants merged)
- NaN preservation (sign-preserving: +nan and -nan not merged)
- Negative zero not merged with positive zero (different bit pattern)
- Complex NaN deduplication
- Complex negative zero preservation
- NaN deduplication via non-factory ops (e.g., `clamp`)
- NaN deduplication via constant pad

Generated by my agent

### CODE REVIEW

**Comment by ezyang** [comment]
@claude is the modified code handling full only or do I have to worry about other functions?

**Comment by ezyang** [comment]
@claude idk, reason about all the other cases, this seems sus.

**Comment by KarhouTam** [comment]
@claude how it looks now

**Comment by KarhouTam** [comment]
@claude how it looks now

**Comment by KarhouTam** [comment]
@claude what about now

**Comment by KarhouTam** [comment]
@claude how it looks now

**Comment by ezyang** [comment]
@claude Two questions: `__CSE_FLOAT__ ` this is a sentinel, which means adversarial inputs could result in confusion. Is it possible to adversarially trigger this? Second, do the tests actually demonstrate the failure without this change or merely exercise the code?

**Comment by ezyang** [comment]
@claude Two questions: __CSE_FLOAT__  this is a sentinel, which means adversarial inputs could result in confusion. Is it possible to adversarially trigger this? Second, do the tests actually demonstrate the failure without this change or merely exercise the code?

**Comment by KarhouTam** [comment]
Hi @ezyang, thanks for your review and valuable feedback. I have addressed your comments. Regarding the CSE pass mentioned by claude, should I include the fix in this PR or handle it in a follow-up issue?

@claude, are these changes safe? Could they introduce any regressions or unhandled actions by now?

**Review by ezyang — APPROVED** [review]

**Review by ezyang — CHANGES_REQUESTED** [review]
beh, sol points out a regression

```
• PR Review: #191173

  ### Findings

  1. Correctness regression: torch/_functorch/compile_utils.py:33 uses
     isinstance, collapsing float/complex subclasses, NumPy scalars, and float
     enums into the same key as built-ins. This defeats the existing type-
     sensitive hash at torch/_functorch/compile_utils.py:239, allowing incorrect
     merges when a callable observes argument type. Use type(val) is float /
     type(val) is complex.

  2. Incomplete normalization: only pytree leaves are normalized. Dictionary
     keys live in TreeSpec context, so {0.0: ...} and {-0.0: ...} can still
     merge, while identical NaN keys fail to merge. This is pre-existing, but
     demonstrates the deeper issue: CSE equivalence is split between flattened
     leaves, TreeSpec, hashing, and token equality.

  3. Test gap: all NaN tests use check_val=False, so the transformed graphs are
     constructed but never executed. torch/testing/_internal/
     common_utils.py:4723 already supports equal_nan=True; the tests should
     verify execution and results. Complex tests also only exercise the real
     component.

  4. PR description is stale: it still describes a NaN-only tuple sentinel,
     while the implementation now bit-keys every built-in float and complex
     value using _ScalarKey, including signed zero behavior.

  ### Recommendation

  Request Changes. The bit-pattern approach is sound for exact built-in scalars,
  but the subclass regression should be fixed before merging. The broader CSE-
  key architecture and second CSE implementation can reasonably be follow-ups,
  provided the PR narrows its claims.
```

**Comment by KarhouTam** [comment]
Hi @ezyang. Except `#3`, other points are fair. `#1` and `#4` look good. `#2` is pre-existing and I am happy to open a follow-up PR to fix that.

>   3. Test gap: all NaN tests use check_val=False, so the transformed graphs are
     constructed but never executed. torch/testing/_internal/
     common_utils.py:4723 already supports equal_nan=True; the tests should
     verify execution and results. Complex tests also only exercise the real
     component.
 
`check()` doesn't use `assertEqual`. It uses `torch.all(true_result == our_result)`, which shall not let NaN results pass. I suggest we refactor the `check()` and related tests also, in a follow-up PR. Wdyt?

**Comment by KarhouTam** [comment]
Hi @ezyang. Sorry to pin you. I've fixed it according to the feedback. Would you please re-review this PR? Thanks!

**Comment by ezyang** [comment]
@claude how does it look now?

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
## mergedog handoff


> [!NOTE]
> This PR is being shepherded by [mergedog](https://github.com/ezyang/mergedog), an automation tool run by a maintainer. It watches CI and may push fixes; the PR author does not need to act unless a maintainer asks.

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.


Current PR head: `57d9a62b93bd2af4bd0704d4c94b0eea2ddc9fca`.
No mergedog-authored commits were pushed during this run.

No LLM was invoked during this run.
